pkg/fieldpath/paved.go

```diff
@@ -0,0 +1,266 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldpath
+
+import (
+	"encoding/json"
+
+	"github.com/pkg/errors"
+)
+
+// A Paved JSON object supports getting and setting values by their field path.
+type Paved struct {
+	object map[string]interface{}
+}
+
+// Pave a JSON object, making it possible to get and set values by field path.
+func Pave(object map[string]interface{}) *Paved {
+	return &Paved{object: object}
+}
+
+// MarshalJSON to the underlying object.
+func (p Paved) MarshalJSON() ([]byte, error) {
+	return json.Marshal(p.object)
+}
+
+// UnmarshalJSON from the underlying object.
+func (p *Paved) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, &p.object)
+}
+
+func (p *Paved) getValue(s Segments) (interface{}, error) {
+	var it interface{} = p.object
+	for i, current := range s {
+		final := i == len(s)-1
+		switch current.Type {
+		case SegmentIndex:
+			array, ok := it.([]interface{})
+			if !ok {
+				return nil, errors.Errorf("%s: not an array", s[:i])
+			}
+			if int(current.Index) >= len(array) {
+				return nil, errors.Errorf("%s: no such element", s[:i+1])
+			}
+			if final {
+				return array[current.Index], nil
+			}
+			it = array[current.Index]
+
+		case SegmentField:
+			object, ok := it.(map[string]interface{})
+			if !ok {
+				return nil, errors.Errorf("%s: not an object", s[:i])
+			}
+			v, ok := object[current.Field]
+			if !ok {
+				return nil, errors.Errorf("%s: no such field", s[:i+1])
+			}
+			if final {
+				return v, nil
+			}
+			it = object[current.Field]
+		}
+	}
+
+	// This should be unreachable.
+	return nil, nil
+}
+
+// GetString value of the supplied field path.
+func (p *Paved) GetString(path string) (string, error) {
+	segments, err := Parse(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "cannot parse path %q", path)
+	}
+
+	v, err := p.getValue(segments)
+	if err != nil {
+		return "", err
+	}
+
+	s, ok := v.(string)
+	if !ok {
+		return "", errors.Errorf("%s: not a string", segments)
+	}
+	return s, nil
+}
+
+// GetBool value of the supplied field path.
+func (p *Paved) GetBool(path string) (bool, error) {
+	segments, err := Parse(path)
+	if err != nil {
+		return false, errors.Wrapf(err, "cannot parse path %q", path)
+	}
+
+	v, err := p.getValue(segments)
+	if err != nil {
+		return false, err
+	}
+
+	b, ok := v.(bool)
+	if !ok {
+		return false, errors.Errorf("%s: not a bool", segments)
+	}
+	return b, nil
+}
+
+// GetNumber value of the supplied field path.
+func (p *Paved) GetNumber(path string) (float64, error) {
```

Comment:
Do you think helpers like `GetNumberAsInt` could be useful here?

Reply:
I thought about that, but given JSON numbers should always unmarshal as a float64 it seemed safer to return the float64 and let the caller handle conversion. This way it will always be obvious to the caller that the field value might have been a float, and thus might be truncated if/when it's asserted to an int.

Reply:
SGTM